rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md

```diff
@@ -0,0 +1,115 @@
+# RFC 9460 - 2021-12-01 - Healthcheck endpoint improvements
+
+Our existing health endpoint is too limited and additional features are required
+to properly interact with load balancers or Kubernetes service discovery.
+
+## Context
+
+- [Additional response for api health endpoint](https://github.com/vectordotdev/vector/issues/9160)
+
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+
+### Out of scope
+
+- Moving the health endpoint out of our existing API
+- Adding user facing configuration to adjust endpoint behavior or responses
+
+## Pain
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on Vector's API is enhanced to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+### Implementation
+
+We have existing logic in place for other components to handle Vector's shutdown.
+The [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L7)
+can be updated to respond 503 if shutdown has started and 200 otherwise. This
```

Comment:
Is there a cause to consider issuing a 503 during start-up, too?

Despite shutdown generally being the longer phase, there are situations were start-up can also be relatively expensive.

Third-party API key validation, using a `[provider]` block to retrieve a remote configuration, an eventual enterprise remote loading strategy, etc. These may be asynchronous concerns per-component, but they can add up to wall clock delays between a Vector process being started and it being ready to ingest events.

We could move HTTP initialization above config validation and issue a 503 until config is parsed, to get a more accurate readiness indicator.

Reply:
I'd think adjusting _when_ the warp server starts is better than having it start and return 503's until "Vector is ready" (which, IMHO isn't well defined/queryable today). We'd still be considered "down" for service discovery mechanisms and achieve the same goal.

Reply:
Currently, warp starts _after_ a configuration has been loaded (and in some cases, retrieved), so if we don't want to distinguish between "Vector has started but is not yet ready to receive/ingest events" and "Vector is ready", that's already the current state.

However, if we don't care about that distinction during start-up, I'd wonder why we care during shutdown. Why not just take `/health` offline entirely to mirror the 'not reachable' state during start-up?

Reply:
That's essentially what the RFC proposes, no? Today during shutdown we're still reachable and still responding as "up".

I think the API can be valuable during shutdown for monitoring purposes and don't think I'd recommend shutting down the API prior to the rest of the components.

Reply:
My main point is that I feel that "startup" and "shutdown" are somewhat equivalent.

They both describe a state in which the Vector _process_ is healthy (in the context of a running binary), but ingestion is inactive or winding down toward inactive.

If we are going to return 503 for one of them, it feels like we should return 503 for the other for completeness.

There _is_ a nuance between both states, which I guess could be an argument against it- during startup, a config could fail to be retrieved or parse, where shutdown _generally_ follows a health config that has been explicitly told to stop. I don't think that's always the case, though (a component can fail and therefore cause a shutdown.), so this still feels somewhat equivalent.

I'm not overly opinionated on this point. It just feels incomplete to accommodate shutdown but not startup.